clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -309,11 +309,14 @@ private boolean renameDirectory(Path src, Path dst) throws IOException {
             dstFileStatus = getFileStatus(dst);
             dstExists = true;
             if (!dstFileStatus.isDirectory()) {
-                //TODO (Tals): move all of this to the input validation part of the rename functionality with https://github.com/treeverse/lakeFS/issues/2061
-                // Mimics the behaviour in https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1527
-                LOG.error("renameDirectory: src {} is a directory and dst {} is a file", src, dst);
-                return false;
+                // Same as https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1527
+                throw new FileAlreadyExistsException("Failed rename " + src + " to " + dst
+                        + ". destination file already exists.");
             }
+            // lakefsFs only has non-empty directories. therefore, is the destination is an existing directory we consider
+            // it to be non-empty. The behaviour is same as https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1530
```

Comment:
I read that code and I don't understand: https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1531 throws a `RenamedFailedException`, here we report an error and fail the operation by returning false.

Reply:
If you look into https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1471 you can see how s3a rename operation handles RenameFailedExceptions. It returns its exit code which is false in case of dir -> non-empty dir rename.

Reply:
Ouch!